contracts/LiquidityConversionRates.sol

```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
+          require(_feeInBps < 10000);
+          feeInBps = _feeInBps;
     }
 
-    function getRateWithE(ERC20 token, uint currentBlockNumber, bool buy, uint qty, uint E) public view returns(uint) {
+    function getRateWithE(ERC20 conversionToken, bool buy, uint qty, uint E) public view returns(uint) {
         uint deltaE;
         uint deltaT;
         uint rate;
+        uint maxCap;
 
-        currentBlockNumber;
+        if(conversionToken != token) return 0;
 
         if(buy) {
           // ETH goes in, token goes out
-          deltaE = qty * precision / 10**ETH_DECIMALS;
-          if(deltaE == 0) return 0;
+          deltaE = fromWeiToQty(qty);
+          if(deltaE == 0) deltaE = 1;
 
-          deltaT = deltaTFunc(r,Pmin,E,deltaE,precision);
+          deltaT = deltaTFunc(r,Pmin,E,deltaE,qtyPrecision);
+          deltaT = reduceFee(deltaT);
 
           rate = deltaT * PRECISION / deltaE;
+
+          maxCap = maxEthCapBuy;
         }
         else {
-          deltaT = qty * precision / (10** getDecimals(token));
-          if(deltaT == 0) return 0;
+          deltaT = fromDecimalsToQty(qty);
+          deltaT = reduceFee(deltaT);
+          if(deltaT == 0) deltaT = 1;
 
-          deltaE = deltaEFunc(r,Pmin,E,deltaT,precision,numPrecisionBits);
+          deltaE = deltaEFunc(r,Pmin,E,deltaT,qtyPrecision,numPrecisionBits);
           rate = deltaE * PRECISION / deltaT;
-        }
 
-        // TODO - check imbalance
+          maxCap = maxEthCapSell;
+        }
 
+        if(deltaE > maxCap) return 0;
         return rate;
     }
 
-    function getRate(ERC20 token, uint currentBlockNumber, bool buy, uint qty) public view returns(uint) {
-        uint E = reserveContract.balance * precision / 10**ETH_DECIMALS;
+    function getRate(ERC20 conversionToken, uint currentBlockNumber, bool buy, uint qty) public view returns(uint) {
+
+        currentBlockNumber;
+
+        uint E = fromWeiToQty(conversionToken.balance);
+        // Yaron - here we need to reduce what was calculated for fee? I think not since we deal with token fee now.
+
+        return getRateWithE(token,buy,qty,E);
+    }
+
+    function recordImbalance(
+        ERC20 conversionToken,
+        int buyAmount,
+        uint rateUpdateBlock,
+        uint currentBlock
+    )
+        public
+    {
+        conversionToken;
+        rateUpdateBlock;
+        currentBlock;
+
+        uint buyQty = fromDecimalsToQty(abs(buyAmount));
+        uint feeQty = getFee(buyQty);
+        collectedTokenFees += feeQty;
+
+        // Yaron - do we need? -require(msg.sender == reserveContract);
+        // Yaron - do we need? - if (rateUpdateBlock == 0) rateUpdateBlock = getRateUpdateBlock(token);
+        // Yaron - do we need it - return addImbalance(token, buyAmount, rateUpdateBlock, currentBlock);
+    }
+
+    function resetCollectedFees() public {
+        collectedTokenFees = 0;
+    }
+
+    function getCollectedFees() public view returns(uint) { //Yaron - Do we need it?
```

Comment:
if you want to have it, call it `getCollectedFeesInTokenWei`.

Reply:
removed it, now wesave in storage collectedFeesInTwei, so no additional value for a dedicated function.